operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+	cilium_client_v2alpha1 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_labels "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_meta_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+const (
+	// The condition added to services to indicate if a request for IPs could be satisfied or not
+	ciliumSvcRequestSatisfiedCondition = "io.cilium/lb-ipam-request-satisfied"
+
+	// The annotation LB IPAM will look for when searching for requested IPs
+	ciliumSvcLBIPSAnnotation = "io.cilium/lb-ipam-ips"
+
+	// The string used in the FieldManager field on update options
+	ciliumFieldManager = "cilium-operator-lb-ipam"
+
+	serviceNamespaceLabel = "io.kubernetes.service.namespace"
+	serviceNameLabel      = "io.kubernetes.service.name"
+)
+
+var Cell = cell.Module(
+	"LB-IPAM",
+	cell.Config(defaultLBIPAMConfig),
+	// Provide LBIPAM so instances of it can be used while testing
+	cell.Provide(registerLBIPAM),
+	// Invoke an empty function which takes an LBIPAM to force its construction.
+	cell.Invoke(func(*LBIPAM) {}),
+)
+
+var defaultLBIPAMConfig = LBIPAMConfig{
+	EnableLBIPAM: false,
+}
+
+type LBIPAMConfig struct {
+	EnableLBIPAM bool `mapstructure:"lb-ipam"`
+}
+
+func (cfg LBIPAMConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("lb-ipam", cfg.EnableLBIPAM, "Enables LB IPAM, which manages IPs for LoadBalancer services internally without cloud providers")
+}
+
+type LBIPAMInitDone chan struct{}
+
+type LBIPAMParams struct {
+	cell.In
+
+	Logger       logrus.FieldLogger
+	LC           hive.Lifecycle
+	ClientSet    k8sClient.Clientset
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	SvcResource  resource.Resource[*core_v1.Service]
+	Config       LBIPAMConfig
+	InitDone     LBIPAMInitDone `optional:"true"`
+}
+
+func registerLBIPAM(params LBIPAMParams) *LBIPAM {
+	if !params.Config.EnableLBIPAM {
+		return nil
+	}
+
+	var lbClasses []string
+	if operatorOption.Config.BGPAnnounceLBIP {
+		lbClasses = append(lbClasses, "io.cilium/bgp-control-plane")
+	}
+
+	lbIPAM := &LBIPAM{
+		Logger:       params.Logger,
+		PoolResource: params.PoolResource,
+		SvcResource:  params.SvcResource,
+		InitDone:     params.InitDone,
+		PoolStore:    NewPoolStore(),
+		RangesStore:  NewRangesStore(),
+		ServiceStore: NewServiceStore(),
+		LBClasses:    lbClasses,
+		IPv4Enabled:  option.Config.IPv4Enabled(),
+		IPv6Enabled:  option.Config.IPv6Enabled(),
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+	ipamShutdown := make(chan struct{})
+
+	params.LC.Append(hive.Hook{
+		OnStart: func(ctx context.Context) error {
+			params.Logger.Info("Starting LB IPAM")
+
+			// Add the clients at start time since the client set doesn't have the clients available during
```

Comment:
Yeah, let's fix `pkg/k8s/client` and just construct the clients at construction rather than start. I did this to "discourage" use of it before it's started, but it's not worth the pain.